frontend/ui/trapper.lua

```diff
@@ -126,6 +126,7 @@ function Trapper:info(text, fast_refresh)
         return true -- not dismissed
     end
 
+    UIManager:preventStandby()
```

Comment:
Not so sure it's to be done here, in Trapper:info().
It's been a while since I wrote this, so I'm not so sure anymore :) but I think Trapper:info() runs for 100ms, and then give control back to the calling code - and it's in that calling code that some possibly long running blocking stuff is happening (like when downloading an EPUB with 100 images, you'll get 100 Trapper:info("Downloading image 4/100...") - but stuff is done in wikipedia.lua.
So, I guess this shouldn't be done in Trapper:info() and Trapper:confirm() - but in the few bits of code that calls them.

But it's fine in the other dismissable... below.

Reply:
> Trapper:info() runs for 100m

Yup, whenever it yields to ui loop is the moment of UI getting stuck in a standby.

> and it's in that calling code that some possibly long running blocking stuff is happening

All of that long running blocking stuff is wrapped in trapper calls in the first place. Stuff like `ffi.sleep` or io.popen doesn't concern us, they can never invoke ui loop that does the actual standby.

With trapper, we don't care about whatever it is happening in background, we care about telling ui loop to not standby whenever we yield control to it, which is why it's fine to do it for each info(). While this indeed creates needless standby bans, it vastly simplifies the code compared to micromanaging this for each trapper end user individually.

The only time you want to micromanage is when the place runs outside of trapper - polling manually via its own `UIManager:scheduleIn()`. That is certainly the case for covers, bgrunner and as well as whole slew of network stuff (calibre, kosync, evernote..). We don't care about the network ones for the time being, as that would mandate a bit more invasive changes down to luasocket to make it robust enough. For now, it's only safe to autostandby with no network, and in the hands of standby driver to ensure that.

Reply:
> All of that long running blocking stuff is wrapped in trapper calls in the first place. Stuff like ffi.sleep or io.popen doesn't concern us, they can never invoke ui loop that does the actual standby.

But being "wrapped in trapper calls" changes what? It's just being in a subroutine (still in the main and single thread) and being able to yield() to go to UI loop for some small period of time.

If the code does trapper.info("Processing...") - which preventStandby, waits 100ms, allosStandby and give control back to the original code - which then does a computation that would take 20s, the PB system will notice no input for the last 5s and will do standby - while the user is still thinking the computation is happening.

Or may be you should preventStandby/allowStandby in Trapper:wrap() (inside pcalled_func) - so it's prevented for as long as the wrapped function is running ?

Or am I missing something ? (it's been a long time - it's impressive how you got that quickly a hold of our code base :)

Reply:
> It's just being in a subroutine (still in the main and single thread) and being able to yield() to go to UI loop for some small period of time.

There's no harm in, however, freely lifting the ban in the context of the coroutine interminently as long there are no yields inbetween (this happens at the boundary between two info()) - you must ban only before you yield.

You can of course do this from the outside, but effect would be the same, while putting significant burden on all trapper clients.

>Or may be you should preventStandby/allowStandby in Trapper:wrap() (inside pcalled_func) - so it's prevented for as long as the wrapped function is running ?

Yes, you mean:
```
function Trapper:wrap(func)
    -- Catch and log any error happening in func (an error happening
    -- in a coroutine just aborts silently the coroutine)
    local pcalled_func = function()
        -- we use xpcall as it can give a whole stacktrace, unlike pcall
        local ok, err = xpcall(function()
            UIManager:preventStandby()
            func()
            UIManager:allowStandby()
        end, debug.traceback)
        if not ok then
            logger.warn("error in wrapped function:", err)
            return false
        end
        return true
        -- As a coroutine, we will return at first coroutine.yield(),
        -- and the above true/false won't probably be caught by
        -- any code, but let's do it anyway.
    end
    local co = coroutine.create(pcalled_func)
    return coroutine.resume(co)
end
```

This is what I've started with, the problem is error(). I'd need to wrap all coroutine.resume() in callbacks with error checks to unwind the ban, making an elegant solution rather messy in practice.

The current approach counts on that we have a good control over error in limited section before each yield, and in wherever else the standby ban is un-bumped so no harm done on error(). Whereas holding the ban for entire duration of trapper that is series of yields, a plugin can mess up, crash the coroutine, and subsequently leave the counter stuck.

Reply:
> I guess this is the source of misunderstanding

Yes, I might not understand how PB standby happens.

> The invariant everywhere is that standby can't just kick in willy nilly even if unbanned. Only a routine called by next UI loop trip (input event hook/driver, looper, timer..) can actually freeze the system

Not sure I understand. Why can't it just kick in?
And which is that routine ? Isn't it the PB system that will just decide to go into standby without knowing KOReader working state ? Or do/will you have to, in our UI loop, do the inkview.allowStanby / sleep+waitforinut / disallowStandby ? May be it's this that I don't get.

> Yes, you mean:

Or rather if you don't want to miss allowStandby in case of error:

```lua
function Trapper:wrap(func)
    -- Catch and log any error happening in func (an error happening
    -- in a coroutine just aborts silently the coroutine)
    local pcalled_func = function()
        UIManager:preventStandby()
        -- we use xpcall as it can give a whole stacktrace, unlike pcall
        local ok, err = xpcall(function()
            func()
        end, debug.traceback)
        UIManager:allowStandby()
        if not ok then
            logger.warn("error in wrapped function:", err)
            return false
        end
        return true
        -- As a coroutine, we will return at first coroutine.yield(),
        -- and the above true/false won't probably be caught by
        -- any code, but let's do it anyway.
    end
    local co = coroutine.create(pcalled_func)
    return coroutine.resume(co)
end
```

Reply:
>waitforinut
In case of pocketbook, it's waitforinput that kicks in standby if it's unbanned at that point. In case of kobo, you'd do the same from your scheduled timer. Effect is the same - both are run from ui loop, *outside* of context of any trapper coroutine - you have to yield into such code path.

```
        if not ok then
            logger.warn("error in wrapped function:", err)
            return false
        end
```

This does nothing on error() past first yield(). I'm not opposed to this approach, *provided* coroutine.resume() in callbacks actually does catch the arrors in some central place. This amounts to more changes than I'm comfortable with.

Reply:
> In case of pocketbook, indeed it's waitforinput that kicks in standby if it's unbanned 

OK, if that's true, I think I'm fine with your current approach :) 
Can you point at which part of the pocketbook code this is allowed ? I don't see anything explicitely mentionning that in https://github.com/koreader/koreader-base/pull/1168.
May be add some comments before/after the explicite inkview call (I guess), Could it be this?

```lua
inkview.ProcessEventLoop() -- standby is allowed when in this
-- standby won't happen after this
```

> This does nothing on error() past first yield(). 

With what did you test it ? The wikipedia Save as EPUB has its own handler for error that will prevent the main Trapper:wrap one to kick in. If I remove it and have the main trapper one handle it, I get after a few Trapper:info():
```
08/25/20-16:04:25 WARN  error in wrapped function: frontend/ui/wikipedia.lua:1352: attempt to index global 'yyy' (a nil value)
stack traceback:
 frontend/ui/wikipedia.lua:1352: in function 'createEpub'
 frontend/ui/wikipedia.lua:1423: in function <frontend/ui/wikipedia.lua:1421>
 [C]: in function 'xpcall'
 frontend/ui/trapper.lua:47: in function <frontend/ui/trapper.lua:45>
```

Reply:
> Ok, I'll force-push this with xpcall global wrapper in a while

Don't feel obliged if it works with that UI loop wait for input granularity.
Might be better if the user falls asleep while on a `Trapper:confirm("Do you want to proceed?") `call :)